go/kbfs/libkbfs/subscription_manager.go

```diff
@@ -0,0 +1,331 @@
+	if err != nil {
+		return "", err
+	}
+	nitp := parsedPath.getNormalizedInTlfPath()
+	sid := makeSubscriptionID()
+
+	ref := pathSubscriptionRef{
+		folderBranch: fb,
+		path:         nitp,
+	}
+
+	sm.lock.Lock()
+	defer sm.lock.Unlock()
+	sm.config.Notifier().RegisterForChanges(
+		[]data.FolderBranch{ref.folderBranch}, sm)
+	if sm.pathSubscriptions[ref] == nil {
+		sm.pathSubscriptions[ref] = make(map[SubscriptionID]debouncedNotify)
+	}
+	limit := rate.Inf
+	if deduplicateInterval != nil {
+		limit = rate.Every(*deduplicateInterval)
+	}
+	sm.pathSubscriptions[ref][sid] = debounce(func() {
+		notifier.OnPathChange(sid, path, topic)
+	}, limit)
+	sm.pathSubscriptionIDToRef[sid] = ref
+	return sid, nil
+}
+
+func (sm *subscriptionManager) subscribeNonPath(ctx context.Context,
+	topic keybase1.SubscriptionTopic, deduplicateInterval *time.Duration,
+	notifier SubscriptionNotifier) (SubscriptionID, error) {
+	sid := makeSubscriptionID()
+
+	sm.lock.Lock()
+	defer sm.lock.Unlock()
+	if sm.nonPathSubscriptions[topic] == nil {
+		sm.nonPathSubscriptions[topic] = make(map[SubscriptionID]debouncedNotify)
+	}
+	limit := rate.Inf
+	if deduplicateInterval != nil {
+		limit = rate.Every(*deduplicateInterval)
+	}
+	sm.nonPathSubscriptions[topic][sid] = debounce(func() {
+		notifier.OnChange(sid, topic)
+	}, limit)
+	sm.nonPathSubscriptionIDToTopic[sid] = topic
+	return sid, nil
+}
+
+func (sm *subscriptionManager) unsubscribePath(
+	ctx context.Context, subscriptionID SubscriptionID) {
+	sm.lock.Lock()
+	defer sm.lock.Unlock()
+	ref, ok := sm.pathSubscriptionIDToRef[subscriptionID]
+	if !ok {
+		return
+	}
+	delete(sm.pathSubscriptionIDToRef, subscriptionID)
+	if (sm.pathSubscriptions[ref]) == nil {
+		return
+	}
+	if notifier, ok := sm.pathSubscriptions[ref][subscriptionID]; ok {
+		notifier.shutdown()
+	}
+	delete(sm.pathSubscriptions[ref], subscriptionID)
+	if len(sm.pathSubscriptions[ref]) == 0 {
+		sm.config.Notifier().UnregisterFromChanges(
+			[]data.FolderBranch{ref.folderBranch}, sm)
+		delete(sm.pathSubscriptions, ref)
+	}
+}
+func (sm *subscriptionManager) unsubscribeNonPath(
+	ctx context.Context, subscriptionID SubscriptionID) {
+	sm.lock.Lock()
+	defer sm.lock.Unlock()
+	topic, ok := sm.nonPathSubscriptionIDToTopic[subscriptionID]
+	if !ok {
+		return
+	}
+	delete(sm.nonPathSubscriptionIDToTopic, subscriptionID)
+	if sm.nonPathSubscriptions[topic] == nil {
+		return
+	}
+	if notifier, ok := sm.nonPathSubscriptions[topic][subscriptionID]; ok {
+		notifier.shutdown()
+	}
+	delete(sm.nonPathSubscriptions[topic], subscriptionID)
+}
+
+func (sm *subscriptionManager) nodeChange(node Node) {
+	sm.lock.RLock()
+	defer sm.lock.RUnlock()
+	path, ok := node.GetPathPlaintextSansTlf()
+	if !ok {
+		return
+	}
+	ref := pathSubscriptionRef{
+		folderBranch: node.GetFolderBranch(),
+		path:         normalizedInTlfPath(path),
+	}
+	if sm.pathSubscriptions[ref] == nil {
+		return
+	}
+	for _, notifier := range sm.pathSubscriptions[ref] {
+		notifier.notify()
```

Comment:
So this is intentionally triggering all topics for the path for now?  For example, a `touch` of a directory will trigger a `NodeChange` event for that directory's node, but the children of the directory will not have changed.  Similarly if someone (mistakenly?) signed up for children notifications on a file, they would get them whenever the file was written to.  Probably not a big deal but maybe worth documenting somewhere?  I'm just trying to figure out why path-based subscriptions have topics at all.

Reply:
Admittedly I didn't put much thoughts on making this exact. Since the notification just triggers a load on the GUI side, I was just going along the path to rather have more notifications than missing one. Ideally it should do exactly as the topics describe, i.e. notifying children when children changes, and notifying stat when the file (or dir) itself changes. Perhaps we can do this in a separate ticket. I'll add a comment.

The reason why I put multiple topics here is not only for optimization (i.e. the "refining" mentioned above), but also so we don't have to keep track of topics in GUI. Then when receiving notification we can just mechanically trigger a load based on the topic inside the notification.